# Visualization Project

---
**Authors**:
-  *Juan P. Zaldivar E.*
-  *Enrique Millán X.*
---

## Introduction

This file contains the preprocessing required for the datasets. The datasets are:
- Colissions dataset.
- Weather dataset.
- New York Map.


<!-- explicar que primer se hace una exploracion visual y luego el preprocesing? -->

### Dataset obtention

The [*collision*](https://data.cityofnewyork.us/Public-Safety/Motor-Vehicle-Collisions-Crashes/h9gi-nx95) dataset was already given by the instructors of the project.

The *weather* dataset was obtained following the next steps:

- Visit the [NOAA Climate Data Online Search](https://www.ncdc.noaa.gov/cdo-web/search) web page.

- Select the following options:
  - `Weather Observation Type/Dataset -> Daily Summaries, Date Range -> 2018-01-01 to 2020-12-31, Search For -> Cities, Search Term -> New York City.`

- Look for "*New York, NY US*" and click in ADD TO CART. Now, click the cart in the top right corner.

- Select "*Custom GHCN-Daily CSV*", and the date previously selected (2018-01-01 to 2020-12-31). We are selecting more information than needed (to avoid disjoint downloads), but we will later filter it with ``Pandas`` and ``Open Refine``. Click continue.

- Fill the three options, and select "*metric units*".

- Fill all the options remaining and click continue. There are some options that will be probably not needed, but we will further analyze this when cleaning the datasets.

- Type the email where you want to receive the data so the order can start.

The *map* dataset was obtained following the next steps:

The datasets are located in the folder `Data/` and the results are saved in the folder `Data/Preprocessed/`. Following are the loading of each dataset and the import of the required libraries.

### Libraries

In [1]:
import os
import numpy as np
import pandas as pd

from Modules import collision_preprocessing as cp

In [2]:
# !pip install numpy
# comentar las versiones de los paquetes

## Dataset preprocessing

The preprocessing of the files was done conjuntly with OpenRefine and the python libraries in order to be able to take advantages of both tools.

In [3]:
dir = './Data'
colission_exists = False

## Collision dataset preprocessing

The original dataset was first loaded into a ``Pandas`` dataframe in order to filter the desired range of dates. The reason lays in a more efficient way to filter the data, taking the size of the original dataset into account. This volumn of data made the computational process in OpenRefine very slow and ineffcient. After this initial filtering, the dataset was exported to a ``.csv`` file and loaded into OpenRefine.

In [4]:
if os.path.exists(f'{dir}/collisions_2018-2020.csv'):
    collision = pd.read_csv(f'{dir}/collisions_2018-2020.csv')
    colission_exists = True
else:
    collision = pd.read_csv(f'{dir}/collisions.csv')

collision.shape

(115740, 29)

In [5]:
collision.head()

CRASH DATE CRASH TIME    BOROUGH  ZIP CODE   LATITUDE  LONGITUDE  \
0  2020-09-06      18:05        NaN       NaN  40.771038  -73.83413   
1  2020-09-20       9:14        NaN       NaN  40.722095  -73.77772   
2  2020-09-24      22:00  MANHATTAN   10039.0  40.824757  -73.94052   
3  2020-08-06       5:30        NaN       NaN        NaN        NaN   
4  2020-08-01      17:16      BRONX   10462.0  40.840508  -73.85515   

                 LOCATION         ON STREET NAME CROSS STREET NAME  \
0  (40.771038, -73.83413)  WHITESTONE EXPRESSWAY               NaN   
1  (40.722095, -73.77772)     GRAND CENTRAL PKWY               NaN   
2  (40.824757, -73.94052)               8 AVENUE   WEST 148 STREET   
3                     NaN                    NaN               NaN   
4  (40.840508, -73.85515)    METROPOLITAN AVENUE      LINDEN DRIVE   

               OFF STREET NAME  ...   CONTRIBUTING FACTOR VEHICLE 2  \
0                          NaN  ...           Following Too Closely   
1                          NaN  ...                     Unspecified   
2                          NaN  ...  Driver Inattention/Distraction   
3  120       Huge Grant Circle  ...                     Unspecified   
4                          NaN  ...                     Unspecified   

   CONTRIBUTING FACTOR VEHICLE 3  CONTRIBUTING FACTOR VEHICLE 4  \
0                            NaN                            NaN   
1                            NaN                            NaN   
2                            NaN                            NaN   
3                            NaN                            NaN   
4                            NaN                            NaN   

   CONTRIBUTING FACTOR VEHICLE 5  COLLISION_ID  \
0                            NaN       4345591   
1                            NaN       4459141   
2                            NaN       4461437   
3                            NaN       4336560   
4                            NaN       4334713   

                   VEHICLE TYPE CODE 1                  VEHICLE TYPE CODE 2  \
0  Station Wagon/Sport Utility Vehicle                           Motorcycle   
1                                Sedan                        Pick-up Truck   
2                                Sedan                         Motorscooter   
3                                Sedan  Station Wagon/Sport Utility Vehicle   
4                                Sedan                               E-Bike   

   VEHICLE TYPE CODE 3 VEHICLE TYPE CODE 4 VEHICLE TYPE CODE 5  
0                  NaN                 NaN                 NaN  
1                  NaN                 NaN                 NaN  
2                  NaN                 NaN                 NaN  
3                  NaN                 NaN                 NaN  
4                  NaN                 NaN                 NaN  

[5 rows x 29 columns]

If the filtered version does not exist, we proceed with the filtering. To filter the data to the summer of 2018 and 2020, the first step is to change the data type of the **CRASH DATE** attribute.

In [6]:
if not colission_exists:
    collision['CRASH DATE'] = pd.to_datetime(collision['CRASH DATE'])

    collision = collision[((collision['CRASH DATE'] >= '2018-06-01') & (collision['CRASH DATE'] <= '2018-09-30')) | ((collision['CRASH DATE'] >= '2020-06-01') & (collision['CRASH DATE'] <= '2020-09-30'))]
    collision.to_csv(f'{dir}/collisions_2018-2020.csv', index=False)

collision.shape

(115740, 29)

### Data exploration?

The collision table incorporates information pertaining to individual crash events, with each row representing a distinct collision incident. The dataframe compile details from all police reported motor vehicle collisions in NYC. 

<!-- ...se hace una breve descripcion de la exploracion? -->

After the filtering, the dataset was exported to a ``.csv`` file and loaded into OpenRefine. The procedure and reasoning taken in OpenRefine will be explained and justified in the present section.

### Data selection

At first glance, **ON STREET NAME** and **OFF STREET NAME** seem to be the same attribute, but with different names. After looking at the description of the attributes in the web site of the dataset, the following descriptions:

- **ON STREET NAME**: *Street on which the collision occurred*.
- **OFF STREET NAME**: *Street address if known*.

Which gives the idea that both attributes contain approximatelly the same information. Furthermore, there are no rows with both attributes filled, which makes the idea of merging both attributes plausible.

In [7]:
collision[(collision['ON STREET NAME'].notnull()) & (collision['OFF STREET NAME'].notnull())].shape

(0, 29)

In [8]:
collision[(collision['ON STREET NAME'].notnull()) | (collision['OFF STREET NAME'].notnull())].shape

(115514, 29)

The resulting attribute after merging both columns is called **STREET NAME** and contains the street name/address where the collision occurred, with no missing values. Some rows will have a more detail description of the street, while others will only have the name of the street. 

<!-- This is not a problem, since the attribute will be used to join the dataset with the weather dataset, and the weather dataset only contains the name of the street. -->

**CROSS STREET**, which was the third attribute related to the street enviroment could be dropped since at first glance does not seem to be useful for the analysis. However, we have decided to keep the attribute for teh time being since it could can contribute some extra information to the final visualization.

Similarly, **LOCATION** seems to contain the tuple (**LATITUDE**, **LONGITUDE**), so we could, a priori, remove the two extra attributes.

In [9]:
collision[(collision['LOCATION'].notnull()) & (collision['LATITUDE'].notnull()) & (collision['LONGITUDE'].notnull())].shape

(108073, 29)

The number of rows where the three attributes are not missing does not cover the total number of rows, but there are no rows where the **LOCATION** attribute is missing and at least one of the other two attributes is not missing.

In [10]:
collision[(collision['LOCATION'].isnull()) & (collision['LATITUDE'].isnull()) & (collision['LONGITUDE'].notnull()) | (collision['LATITUDE'].notnull()) & (collision['LONGITUDE'].isnull())].shape

(0, 29)

In [11]:
collision[(collision['LOCATION'].isnull())].shape

(7667, 29)

Which makes the rest of the rows (7667) with missing values in the three attributes. This means that the **LATITUDE** and **LONGITUDE** attributes can be removed, since the **LOCATION** attribute contains the same information. By this the number of attributes is reduced by two.

### Data type conversion

se ha realizado un cluster (con el metodo *key collision* y funcion de keying *fingerprint*) de STREET NAME y se han unido las celdas con nombres semejantes.

Luego se han eliminado los espacios sobrantes de la columna STREET NAME.

---

---

### a

In [12]:
precollision = pd.read_csv(f'{dir}/collisions_2018-2020_prepro_v1.csv')
precollision.head()

COLLISION_ID            CRASH DATE CRASH TIME    BOROUGH  ZIP CODE  \
0       4345591  2020-09-06T00:00:00Z      18:05        NaN       NaN   
1       4459141  2020-09-20T00:00:00Z       9:14        NaN       NaN   
2       4461437  2020-09-24T00:00:00Z      22:00  MANHATTAN   10039.0   
3       4336560  2020-08-06T00:00:00Z       5:30        NaN       NaN   
4       4334713  2020-08-01T00:00:00Z      17:16      BRONX   10462.0   

                 LOCATION            STREET NAME CROSS STREET NAME  \
0  (40.771038, -73.83413)  WHITESTONE EXPRESSWAY               NaN   
1  (40.722095, -73.77772)     GRAND CENTRAL PKWY               NaN   
2  (40.824757, -73.94052)               8 AVENUE   WEST 148 STREET   
3                     NaN  120 HUGE GRANT CIRCLE               NaN   
4  (40.840508, -73.85515)    METROPOLITAN AVENUE      LINDEN DRIVE   

   NUMBER OF PERSONS INJURED  NUMBER OF PERSONS KILLED  ...  \
0                        0.0                       1.0  ...   
1                        0.0                       0.0  ...   
2                        0.0                       0.0  ...   
3                        1.0                       0.0  ...   
4                        0.0                       1.0  ...   

                       CONTRIBUTING FACTOR VEHICLE 1  \
0                               Unsafe Lane Changing   
1                     Driver Inattention/Distraction   
2  Pedestrian/Bicyclist/Other Pedestrian Error/Co...   
3                                        Unspecified   
4                      Failure to Yield Right-of-Way   

    CONTRIBUTING FACTOR VEHICLE 2  CONTRIBUTING FACTOR VEHICLE 3  \
0           Following Too Closely                            NaN   
1                     Unspecified                            NaN   
2  Driver Inattention/Distraction                            NaN   
3                     Unspecified                            NaN   
4                     Unspecified                            NaN   

   CONTRIBUTING FACTOR VEHICLE 4  CONTRIBUTING FACTOR VEHICLE 5  \
0                            NaN                            NaN   
1                            NaN                            NaN   
2                            NaN                            NaN   
3                            NaN                            NaN   
4                            NaN                            NaN   

                   VEHICLE TYPE CODE 1                  VEHICLE TYPE CODE 2  \
0  STATION WAGON/SPORT UTILITY VEHICLE                           MOTORCYCLE   
1                                SEDAN                        PICK-UP TRUCK   
2                                SEDAN                         MOTORSCOOTER   
3                                SEDAN  STATION WAGON/SPORT UTILITY VEHICLE   
4                                SEDAN                               E-BIKE   

  VEHICLE TYPE CODE 3 VEHICLE TYPE CODE 4 VEHICLE TYPE CODE 5  
0                 NaN                 NaN                 NaN  
1                 NaN                 NaN                 NaN  
2                 NaN                 NaN                 NaN  
3                 NaN                 NaN                 NaN  
4                 NaN                 NaN                 NaN  

[5 rows x 26 columns]

### Missing values

It has already been mentioned the existence of some missing values. In the previous section, the verification of missing values was done with the ``.isnull()`` method of ``Pandas``. However, this method does not take into account the ``NaN`` values. In order to check the existence of ``NaN`` values, the ``.isna()`` method was used.

In [13]:
comp = (precollision.isnull().sum() == precollision.isna().sum())
comp[comp == False]

KeyboardInterrupt: 

In [ ]:
precollision.isnull().sum()

COLLISION_ID                          0
CRASH DATE                            0
CRASH TIME                            0
BOROUGH                           40671
ZIP CODE                          40686
LOCATION                           7667
STREET NAME                         226
CROSS STREET NAME                 60179
NUMBER OF PERSONS INJURED             2
NUMBER OF PERSONS KILLED              4
NUMBER OF PEDESTRIANS INJURED         0
NUMBER OF PEDESTRIANS KILLED          0
NUMBER OF CYCLIST INJURED             0
NUMBER OF CYCLIST KILLED              0
NUMBER OF MOTORIST INJURED            0
NUMBER OF MOTORIST KILLED             0
CONTRIBUTING FACTOR VEHICLE 1       351
CONTRIBUTING FACTOR VEHICLE 2     19593
CONTRIBUTING FACTOR VEHICLE 3    106611
CONTRIBUTING FACTOR VEHICLE 4    113552
CONTRIBUTING FACTOR VEHICLE 5    115126
VEHICLE TYPE CODE 1                 730
VEHICLE TYPE CODE 2               27447
VEHICLE TYPE CODE 3              107096
VEHICLE TYPE CODE 4              113658


#### Derivation of geographic attributes

#### Derivation of vehicle attributes

In [ ]:
# poner unespecified en factor $i$ si el vehiculo $i$ es null y factor $i$ no es null
# o
# poner unknown en vehiculo $i$ si el vehiculo $i$ es null y factor $i$ no es null

# son disjuntas estas operaciones?????????????

In some rows of the dataset, the **CONTRIBUTING FACTOR VEHICLE** is missing but the **VEHICLE TYPE CODE** is not. This suggests that the vehicle type is known, but the factor that contributed to the collision is not. In order to fill this missing values, the factor was set as *unespecified*. This was done for all the rows and columns where the **CONTRIBUTING FACTOR VEHICLE** was missing with the above condition.

In [ ]:
cp.imputation_with_ref_col(precollision, 'CONTRIBUTING FACTOR VEHICLE', 'VEHICLE TYPE CODE', 'Unspecified', 1, 5)

Likewise, in some rows of the dataset, the **VEHICLE TYPE CODE** is missing but the **CONTRIBUTING FACTOR VEHICLE** is not. This suggests that the factor that contributed to the collision is known, but the vehicle type is not. In order to fill this missing values, the vehicle type was set as *unknown*. This was done for all the rows and columns where the **VEHICLE TYPE CODE** was missing with the above condition.

In [ ]:
cp.imputation_with_ref_col(precollision, 'VEHICLE TYPE CODE', 'CONTRIBUTING FACTOR VEHICLE', 'UNKNOWN', 1, 5)

In [ ]:
precollision.isnull().sum()

COLLISION_ID                          0
CRASH DATE                            0
CRASH TIME                            0
BOROUGH                           40671
ZIP CODE                          40686
LOCATION                           7667
STREET NAME                         226
CROSS STREET NAME                 60179
NUMBER OF PERSONS INJURED             2
NUMBER OF PERSONS KILLED              4
NUMBER OF PEDESTRIANS INJURED         0
NUMBER OF PEDESTRIANS KILLED          0
NUMBER OF CYCLIST INJURED             0
NUMBER OF CYCLIST KILLED              0
NUMBER OF MOTORIST INJURED            0
NUMBER OF MOTORIST KILLED             0
CONTRIBUTING FACTOR VEHICLE 1       288
CONTRIBUTING FACTOR VEHICLE 2     18852
CONTRIBUTING FACTOR VEHICLE 3    106550
CONTRIBUTING FACTOR VEHICLE 4    113540
CONTRIBUTING FACTOR VEHICLE 5    115123
VEHICLE TYPE CODE 1                 288
VEHICLE TYPE CODE 2               18852
VEHICLE TYPE CODE 3              106550
VEHICLE TYPE CODE 4              113540


<!-- comentar la reduccion de missing values -->

Notice that the only missing values in the **CONTRIBUTING FACTOR VEHICLE** attribute and **VEHICLE TYPE CODE** attribute are in the same rows. This means that the number of missing values in the **CONTRIBUTING FACTOR VEHICLE** attribute and **VEHICLE TYPE CODE** attribute is the same.

In [ ]:
# como se hace para imputar los missing de los dos campos a la vez?

#### Derivation of number of person attributes

In [ ]:
# como se hace para imputar los missing del numero de personas? o mejor se elimina?

## Weather dataset preprocessing

Since in the process of gathering the data we selected all possible attributes, now it's time to delete the ones that are completely empty or are not useful for our purpose. Then, other preprocessing steps will be applied.

In [38]:
weather = pd.read_csv(f'{dir}/weather.csv')
weather.shape

D:\USUARIOS\Enric-UPC\AppData\Local\Temp\ipykernel_15012\941678255.py:1: DtypeWarning: Columns (7,9,13,17,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63,65) have mixed types. Specify dtype option on import or set low_memory=False.
  weather = pd.read_csv(f'{dir}/weather.csv')


(90591, 66)

In [39]:
weather.head()

STATION                 NAME  LATITUDE  LONGITUDE  ELEVATION  \
0  USC00280907  BOONTON 1 SE, NJ US  40.89174  -74.39635       85.3   
1  USC00280907  BOONTON 1 SE, NJ US  40.89174  -74.39635       85.3   
2  USC00280907  BOONTON 1 SE, NJ US  40.89174  -74.39635       85.3   
3  USC00280907  BOONTON 1 SE, NJ US  40.89174  -74.39635       85.3   
4  USC00280907  BOONTON 1 SE, NJ US  40.89174  -74.39635       85.3   

         DATE  AWND AWND_ATTRIBUTES  DAPR DAPR_ATTRIBUTES  ...  WT05  \
0  2018-01-01   NaN             NaN   NaN             NaN  ...   NaN   
1  2018-01-02   NaN             NaN   NaN             NaN  ...   NaN   
2  2018-01-03   NaN             NaN   NaN             NaN  ...   NaN   
3  2018-01-04   NaN             NaN   NaN             NaN  ...   NaN   
4  2018-01-05   NaN             NaN   NaN             NaN  ...   NaN   

   WT05_ATTRIBUTES  WT06 WT06_ATTRIBUTES  WT08  WT08_ATTRIBUTES  WT09  \
0              NaN   NaN             NaN   NaN              NaN   NaN   
1              NaN   NaN             NaN   NaN              NaN   NaN   
2              NaN   NaN             NaN   NaN              NaN   NaN   
3              NaN   NaN             NaN   NaN              NaN   NaN   
4              NaN   NaN             NaN   NaN              NaN   NaN   

  WT09_ATTRIBUTES  WT11 WT11_ATTRIBUTES  
0             NaN   NaN             NaN  
1             NaN   NaN             NaN  
2             NaN   NaN             NaN  
3             NaN   NaN             NaN  
4             NaN   NaN             NaN  

[5 rows x 66 columns]

With a simple view we see a lot of very sparse attributes, and a lot of attributes taht are not really useful. To deal with these attributes, the best option is to define the useful ones and erase all the others.

First of all, we will start by joining *LATITUDE* and *LONGITUDE* in a *LOCATION* attribute following the format of the previous dataset. We will delete the *ELEVATION* since it can't be related in any way with the previous dataset and the information it gives is not useful for our purpose.

In [40]:
weather['LOCATION'] = '('+weather['LATITUDE'].astype(str) + ', ' + weather['LONGITUDE'].astype(str)+')'
weather.drop(columns=['LATITUDE', 'LONGITUDE', 'ELEVATION'], inplace=True)

Since we have *LOCATION*, we don't need *STATION* code or *NAME*.

In [41]:
weather.drop(columns=['STATION', 'NAME'], inplace=True)

Now, with regard to the other columns, let's follow the proposed startegy. We are interested in *LOCATION* so we can know about the weather condtitions of the zone where the collision was produced. Now, for the weather attributes, ollowing the *Documentation* of the weather datset, provided by the *NOAA* and available in the Data directory, we are (for now) interested in:

- *PRCP* : Precipitation (mm)
- *SNOW* : Snowfall (mm)
- *SNWD* : Snow depth (mm)
- *TMAX* : Maximum temperature (Celsius)
- *TMIN* : Minimum temperature (Celsius)
- *TOBS* : Temperature at the time of observation
- *AWND* : Average daily wind speed (meters per second)
- *FRTH* : Thickness of frozen ground layer (cm or inches as per user preference)



DATE                object
AWND               float64
AWND_ATTRIBUTES     object
DAPR               float64
DAPR_ATTRIBUTES     object
DASF               float64
DASF_ATTRIBUTES    float64
MDPR               float64
MDPR_ATTRIBUTES     object
MDSF               float64
MDSF_ATTRIBUTES    float64
PGTM               float64
PGTM_ATTRIBUTES     object
PRCP               float64
PRCP_ATTRIBUTES     object
PSUN               float64
PSUN_ATTRIBUTES     object
SNOW               float64
SNOW_ATTRIBUTES     object
SNWD               float64
SNWD_ATTRIBUTES     object
TAVG               float64
TAVG_ATTRIBUTES     object
TMAX               float64
TMAX_ATTRIBUTES     object
TMIN               float64
TMIN_ATTRIBUTES     object
TOBS               float64
TOBS_ATTRIBUTES     object
TSUN               float64
TSUN_ATTRIBUTES     object
WDF2               float64
WDF2_ATTRIBUTES     object
WDF5               float64
WDF5_ATTRIBUTES     object
WESD               float64
WESD_ATTRIBUTES     object
W